WIP: Issue#206 Prepapre API design for Manage flats on /houses#160

tex1988 opened the pull request:
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/206)


## Code reviewers

- [ ] @@kh0ma, @VadymSokorenko, @denis-litvinov, @likeRewca, @Abhai2016, @Ilya-Ross, @annaprok

### Second Level Review

- [ ] @github_username

## Summary of issue

Miss of API describtion for tasks connected with Manage flats on /houses (#180, #181, #182, #183)

## Summary of change

Described API for:
- Getting all apartments from house
- Creating apartment in house
- Updating apartment from house by ID
- Deleting apartment from house by ID
- Getting all invitations for owners of apartment
- Creating invitation for owner of apartment
- Updating invitation for owner of apartment by ID

WARNING: This pull request cannot be merged. It have to be a base for further work on tasks #180, #181, #182, #183

## Testing approach

ToDo

## CHECK LIST
- [x]  СI passed
- [x]  Сode coverage >=95%
- [x]  PR is reviewed manually again (to make sure you have 100% ready code)
- [x]  All reviewers agreed to merge the PR
- [x]  I've checked new feature as logged in and logged out user if needed
- [x]  PR meets all conventions

tex1988 committed:
Issue#206: Changes in openapi.yaml
Added API description for:
- getting all apartments from house
- creating apartment in house
- updating apartment from house by ID
- deleting apartment from house by ID
- getting all invitations for owners of apartment
- creating invitation for owner of apartment
- updating invitation for owner of apartment by ID

tex1988 committed:
Merge branch 'dev' into Issue#206-Prepapre-API-design-for-Manage-flats-on-houses

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -23,6 +23,10 @@ tags:
+  - name: apartment
+    description: Operations with apartment
+  - name: apartment invitation
+    description: Operations with apartment invatations
```
Invitation*
And I didn't get what are the operations with apartment invitations? I mean why do we need them?

tex1988 replied:
Proposed changes have been made

tex1988 replied:
> Invitation*
> And I didn't get what are the operations with apartment invitations? I mean why do we need them?

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      parameters:
+        - name: "house_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the House, please give more understandable descriptions and examples.

tex1988 replied:
> The ID of the House, please give more understandable descriptions and examples.

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+        required: true
+      responses:
+        '201':
+          description: Successfully created
```
As I remember we agreed to use just Success.

tex1988 replied:
> As I remember we agreed to use just Success.
Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    post:
+      tags:
+        - apartment
+      summary: "Create apartment"
```
You can write summary without double quote.

tex1988 replied:
> You can write summary without double quote.

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment
+      summary: "Create apartment"
+      description: Create new Apartment in the house
```
In description we use 'The endpoint does something...' so here it will be - The endpoint creates a new Apartment in the house

tex1988 replied:
> In description we use 'The endpoint does something...' so here it will be - The endpoint creates a new Apartment in the house

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            format: int64
+            example: 1
+      requestBody:
+        description: Create a new Apartment
```
We use 'A payload for...' so it has to be 'A payload for creating a new Apartment'. Check our openapi.yml file to see which patterns we use and follow them please.

tex1988 replied:
> We use 'A payload for...' so it has to be 'A payload for creating a new Apartment'. Check our openapi.yml file to see which patterns we use and follow them please.

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    get:
+      tags:
+        - apartment
+      summary: "Get all apartments"
```
Same, no quotes

tex1988 replied:
> Same, no quotes

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment
+      summary: "Get all apartments"
+      description: Get all apartments of the house
```
Same about endpoint

tex1988 replied:
> Same about endpoint

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          example: 73.5
+      responses:
+        '200':
+          description: Successful operation
```
Success

tex1988 replied:
> Success

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    get:
+      tags:
+        - apartment
+      summary: "Get apartment"
```
Again no quotes

tex1988 replied:
> Again no quotes

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment
+      summary: "Get apartment"
+      description: Get an apartment by an apartment id
```
Again with endpoint

tex1988 replied:
> Again with endpoint

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      parameters:
+        - name: "house_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the house

tex1988 replied:
> The ID of the house

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          example: 1
+        - name: "id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the apartment

tex1988 replied:
> The ID of the apartment

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          example: 1
+      responses:
+        '200':
+          description: Successful operation
```
Success

tex1988 replied:
> Success

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    put:
+      tags:
+        - apartment
+      summary: "Update apartment"
```
Again no quotes

tex1988 replied:
> Again no quotes

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment
+      summary: "Update apartment"
+      description: Update an apartment by apartment ID
```
The endpoint ...

tex1988 replied:
> The endpoint ...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      parameters:
+        - name: "house_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the House

tex1988 replied:
> The ID of the House

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          example: 1
+        - name: "id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the Apartment

tex1988 replied:
> The ID of the Apartment

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            format: int64
+          example: 1
+      requestBody:
+        description: Update a new Apartment
```
A payload....

tex1988 replied:
> A payload....

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+        required: true
+      responses:
+        '200':
+          description: Successful operation
```
Success

tex1988 replied:
> Success

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    delete:
+      tags:
+        - apartment
+      summary: "Delete apartment"
```
Quotes...

tex1988 replied:
> Quotes...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment
+      summary: "Delete apartment"
+      description: Delete an apartment by an apartment ID
```
The endpoint...

tex1988 replied:
> The endpoint...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      parameters:
+        - name: "house_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the House

tex1988 replied:
> The ID of the House

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            example: 1
+        - name: "id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the Apartment

tex1988 replied:
> The ID of the Apartment

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            example: 1
+      responses:
+        "204":
+          description: Successful operation
```
Success

tex1988 replied:
> Success

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    post:
+      tags:
+        - apartment invitation
+      summary: "Create invitation"
```
Quotes...

tex1988 replied:
> Quotes...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment invitation
+      summary: "Create invitation"
+      description: Create an invitation for an apartment owner
```
The endpoint...

tex1988 replied:
> The endpoint...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      parameters:
+        - name: "apartment_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the Apartment

tex1988 replied:
> The ID of the Apartment

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            format: int64
+            example: 1
+      requestBody:
+        description: Create a new Invitation
```
A payload...

tex1988 replied:
> A payload...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+        required: true
+      responses:
+        '201':
+          description: Successfully created
```
Success

tex1988 replied:
> Success

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    get:
+      tags:
+        - apartment invitation
+      summary: "Get all invitations"
```
Quotes...

tex1988 replied:
> Quotes...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment invitation
+      summary: "Get all invitations"
+      description: Get all invitations for apartment owners
```
The endpoin...

tex1988 replied:
> The endpoin...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          example: 1
+      responses:
+        '200':
+          description: Successful operation
```
Success

tex1988 replied:
> Success

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+    patch:
+      tags:
+        - apartment invitation
+      summary: "Update invitation"
```
Quotes...

tex1988 replied:
> Quotes...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      tags:
+        - apartment invitation
+      summary: "Update invitation"
+      description: Update an invitation for an apartment owner by ID
```
The endpoint...

tex1988 replied:
> The endpoint...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+      parameters:
+        - name: "apartment_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the Apartment

tex1988 replied:
> The ID of the Apartment

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            example: 1
+        - name: "id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the invitation

tex1988 replied:
> The ID of the invitation

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+            format: int64
+            example: 1
+      requestBody:
+        description: Add an owner of the apartment
```
A payload...

tex1988 replied:
> A payload...

Proposed changes have been made

Abhai2016 commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+        required: true
+      responses:
+        '200':
+          description: Successful operation
```
Success

tex1988 replied:
> Success

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          $ref: '#/components/responses/Unauthorized'
+        "403":
+          $ref: '#/components/responses/Forbidden'
+        default:
```
What is about 404? I think it can appear here.

tex1988 replied:
> What is about 404? I think it can appear here.

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          $ref: '#/components/responses/BadRequest'
+        "401":
+          $ref: '#/components/responses/Unauthorized'
+        default:
```
The same, 404 could appear here.

tex1988 replied:
> The same, 404 could appear here.

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1107,6 +1111,410 @@ paths:
+          $ref: '#/components/responses/InternalServerError'
+
+  /apartments/{apartment_id}/invitations/{id}:
+    patch:
```
We are not going to implement nor patch nor put for an invitation. Only getOne and delete should be here. Delete can be accomplished only in allowed statuses.

Also we need to have an ability to manage owners - see them, delete, update (in terms of ownership area) but no create.

tex1988 replied:
> We are not going to implement nor patch nor put for an invitation. Only getOne and delete should be here. Delete can be accomplished only in allowed statuses.
> 
> Also we need to have an ability to manage owners - see them, delete, update (in terms of ownership area) but no create.

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1627,3 +2035,244 @@ components:
+    UpdateInvitation:
+      type: object
+      properties:
+        status:
```
status is readonly

tex1988 replied:
> status is readonly

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1627,3 +2035,244 @@ components:
+          multipleOf: 1e-4
+          example: 0.6588
+
+    UpdateInvitation:
```
I think that email also can be updated if the status is pending.

tex1988 replied:
> I think that email also can be updated if the status is pending.

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1627,3 +2035,244 @@ components:
+          maximum: 1.0
+          multipleOf: 1e-4
+          example: 0.6588
+        status:
```
Do we really need it? I think we can go without it.

tex1988 replied:
> Do we really need it? I think we can go without it.

Proposed changes have been made

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -1627,3 +2035,244 @@ components:
+                    main: false
+              ownership_part: 0.38
+              status: active
+        invitations:
```
Let's do not expose invitations as part of `ReadApartment`. We will use a sub-endpoint for this.

tex1988 replied:
> Let's do not expose invitations as part of `ReadApartment`. We will use a sub-endpoint for this.

Proposed changes have been made

tex1988 committed:
Issue#206: Changes in openapi.yaml
- added response variant with 404 status code for query endpoints
- deleted invitations from ReadApartment model
- deleted status from Ownership models
- added ownership endpoint
- changed update endpoint for invitations
- description changes and fixes

tex1988 committed:
Issue#206: Changes in openapi.yaml
- Fix spelling errors
